pkgs/tools/misc/pre-commit/meta-hook-rewiring.patch

```diff
@@ -0,0 +1,182 @@
+From d83e1404b1cf9439786f3c773fb4cdb773d2a11b Mon Sep 17 00:00:00 2001
+From: Matt Rixman <MatrixManAtYrService@users.noreply.github.com>
+Date: Sat, 1 Jul 2023 15:43:37 -0600
+Subject: [PATCH] nixpkgs fixes
+
+---
+ .pre-commit-hooks.yaml            | 21 +++++++++++++++++
+ pre_commit/clientlib.py           | 38 ++++++++++++++++++++++++++++++-
+ pre_commit/commands/autoupdate.py |  5 +++-
+ pre_commit/commands/gc.py         |  3 ++-
+ setup.cfg                         |  2 ++
+ 5 files changed, 66 insertions(+), 3 deletions(-)
+
+diff --git a/.pre-commit-hooks.yaml b/.pre-commit-hooks.yaml
+index e1aaf58..aee247a 100644
+--- a/.pre-commit-hooks.yaml
++++ b/.pre-commit-hooks.yaml
+@@ -4,3 +4,24 @@
+     entry: pre-commit validate-manifest
+     language: python
+     files: ^\.pre-commit-hooks\.yaml$
++-   id: check-hooks-apply
++    name: Check Hooks Apply
++    description: Fails if there are hooks configured that don't match any files in this repo
++    entry: check-hooks-apply
++    language: python
++    types: [python]
++    files: ^\.pre-commit-hooks\.yaml$
++-   id: check-useless-excludes
++    name: Check Useless Excludes
++    description: Fails if there are any exclude expressions which aren't actually excluding anything
++    entry: check-useless-excludes
++    language: python
++    types: [python]
++    files: ^\.pre-commit-hooks\.yaml$
++-   id: identity
++    name: Identity
++    description: Does nothing (exists for testing)
++    entry: check-useless-excludes
++    language: python
++    types: [python]
++    files: ^\.pre-commit-hooks\.yaml$
+diff --git a/pre_commit/clientlib.py b/pre_commit/clientlib.py
+index d0651ca..fbf305f 100644
+--- a/pre_commit/clientlib.py
++++ b/pre_commit/clientlib.py
+@@ -4,7 +4,9 @@ import functools
+ import logging
+ import re
+ import shlex
++import subprocess
+ import sys
++from pathlib import Path
+ from typing import Any
+ from typing import NamedTuple
+ from typing import Sequence
+@@ -146,6 +148,7 @@ load_manifest = functools.partial(
+ 
+ LOCAL = 'local'
+ META = 'meta'
++REPO_ROOT = 'PATH_TO_CLONED_PRECOMMIT'
+ 
+ 
+ class WarnMutableRev(cfgv.Conditional):
+@@ -389,9 +392,42 @@ class InvalidConfigError(FatalError):
+     pass
+ 
+ 
+-load_config = functools.partial(
++load_config_orig = functools.partial(
+     cfgv.load_from_filename,
+     schema=CONFIG_SCHEMA,
+     load_strategy=yaml_load,
+     exc_tp=InvalidConfigError,
+ )
++
++
++Config = dict[str, list[dict[str, Any]]]
++Repo = dict[str, str | list[dict[str, str]]]
++
++
++def load_config(path: str) -> Config:
++    on_disk = load_config_orig(path)
++    modified = replace_meta(on_disk)
++    return modified
++
++
++local_repo = {'repo': REPO_ROOT, 'rev': 'HEAD'}
++
++def replace_meta(config_in: Config) -> Config:
```

Comment:
Typically, a user can just use the string "meta" to indicate a meta hook.  That string is then used by pre-commit to avoid certain steps that would be necessary for an external hook.

Here I modify the way pre-commit reads config from disk so that "meta" is replaced by the contents of the `REPO_ROOT` variable.  I've tested this with `REPO_ROOT` set to `/home/matt/src/pre-commit` (where I had cloned pre-commit by hand) and as long as that repo has these changes, it works.

The goal is to provide a derivation which includes these changes in `/nix/store` so that pre-commit always sees meta hooks as external repositories.  Except "external" in this case just means that they're provided by the pre-commit repo that lives in the nix store.